mpisppy/phbase.py

```diff
@@ -164,25 +164,29 @@ def Compute_Xbar(self, verbose=False, synchronizer=None):
                     if synchronizer is not None and node.name == "ROOT":
                         mylen += self.n_proc
                     local_concats["OnlyReduce"][ndn] = np.zeros(mylen, dtype='d')
-                    node_concats["OnlyReduce"][ndn] = np.zeros(mylen, dtype='d')
+                    global_concats["OnlyReduce"][ndn] = np.zeros(mylen, dtype='d')
 
         # compute the local xbar and sqbar (put the sq in the 2nd 1/2 of concat)
         for k,s in self.local_scenarios.items():
             nlens = s._PySP_nlens        
             for node in s._PySPnode_list:
                 ndn = node.name
-                for i, v in enumerate(node.nonant_vardata_list):
-                    local_concats["OnlyReduce"][ndn][i] += \
-                        (s.PySP_prob / node.uncond_prob) * v._value
-                    local_concats["OnlyReduce"][ndn][nlens[ndn]+i] += \
-                        (s.PySP_prob / node.uncond_prob) * v._value * v._value
+                nlen = nlens[ndn]
+
+                xbars = local_concats["OnlyReduce"][ndn][:nlen]
+                xsqbars = local_concats["OnlyReduce"][ndn][nlen:]
+
+                nonants_array = np.fromiter( (v._value for v in node.nonant_vardata_list),
+                                             dtype='d', count=nlen )
+                xbars += (s.PySP_prob / node.uncond_prob) * nonants_array
+                xsqbars += (s.PySP_prob / node.uncond_prob) * nonants_array**2
```

Comment:
The way this is re-factored you could drop in `s._PySP_Prob_Mask[ndn]`/`s._PySP_Prob_Coeff[ndn]` (as a numpy array) for `(s.PySP_prob / node.uncond_prob)`.

In the typical case (same variable probabilities), `s._PySP_Prob_Coeff[ndn]` could just be set to `(s.PySP_prob / node.uncond_prob)` (and not use an array).